app.js

```diff
@@ -344,64 +395,57 @@ module.exports = {
-        }
-        // Didn't exist, or another fatal error, exit
-        callback(err);
-        return;
-      }
-
-      restoreInternal(
-        restoreDB,
-        opts,
-        srcStream,
-        ee,
-        function(err, writer) {
-          if (err) {
-            callback(err, null);
-            return;
-          }
-          if (writer != null) {
-            addEventListener(listenerErrorIndicator, writer, 'restored', function(obj) {
-              debug(' restored ', obj.total);
-              ee.emit('restored', { documents: obj.documents, total: obj.total });
-            });
-            addEventListener(listenerErrorIndicator, writer, 'error', function(err) {
-              debug('Error ' + JSON.stringify(err));
-              // Only call destroy if it is available on the stream
-              if (srcStream.destroy && srcStream.destroy instanceof Function) {
-                srcStream.destroy();
-              }
-              callback(err);
-            });
-            addEventListener(listenerErrorIndicator, writer, 'finished', function(obj) {
-              debug('restore complete');
-              ee.emit('finished', { total: obj.total });
-              callback(null, obj);
-            });
+    validateArgs(targetUrl, opts)
+      // Set up the DB client
+      .then(() => {
+        opts = Object.assign({}, defaults(), opts);
+        return request.client(targetUrl, opts);
+      })
+      // Validate the DB exists, before proceeding to restore
+      .then((restoreDB) => validateRestoreDb(restoreDB))
+      .then((restoreDB) => {
+        const listenerErrorIndicator = { errored: false };
+        restoreInternal(
+          restoreDB,
+          opts,
+          srcStream,
+          ee,
+          function(err, writer) {
+            if (err) {
+              callback(err, null);
+              return;
+            }
+            if (writer != null) {
+              addEventListener(listenerErrorIndicator, writer, 'restored', function(obj) {
+                debug(' restored ', obj.total);
+                ee.emit('restored', { documents: obj.documents, total: obj.total });
+              });
+              addEventListener(listenerErrorIndicator, writer, 'error', function(err) {
+                debug('Error ' + JSON.stringify(err));
+                // Only call destroy if it is available on the stream
+                if (srcStream.destroy && srcStream.destroy instanceof Function) {
+                  srcStream.destroy();
+                }
+                callback(err);
+              });
+              addEventListener(listenerErrorIndicator, writer, 'finished', function(obj) {
+                debug('restore complete');
+                ee.emit('finished', { total: obj.total });
+                callback(null, obj);
+              });
+            }
           }
-        }
-      );
-    });
+        );
+      })
+      .catch(callback);
```

Comment:
as above

Reply:
Done in https://github.com/IBM/couchbackup/pull/664/commits/5d685ffb5a403c97b91d557d9d1ca1bb3f8013b9